x/smart/racer.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smart
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+// Returns a read channel that is already closed.
+func newClosedChanel() <-chan struct{} {
+	ch := make(chan struct{})
+	close(ch)
+	return ch
+}
+
+// raceTests races will call the test function on each entry until it finds an entry for which the test returns nil.
+// That entry is returned. A test is only started after the previous test finished or maxWait is done, whichever
+// happens first. That way you bound the wait for a test, and they may overlap.
+func raceTests[E any, R any](ctx context.Context, maxWait time.Duration, entries []E, test func(entry E) (R, error)) (R, error) {
+	type testResult struct {
+		Result R
+		Err    error
+	}
+	// Communicates the result of each test.
+	resultChan := make(chan testResult, len(entries))
+	waitCh := newClosedChanel()
+
+	next := 0
+	for toTest := len(entries); toTest > 0; {
+		select {
+		// Search cancelled, quit.
+		case <-ctx.Done():
+			var empty R
+			return empty, ctx.Err()
+
+		// Ready to start testing another resolver.
+		case <-waitCh:
+			entry := entries[next]
+			next++
+
+			waitCtx, waitDone := context.WithTimeout(ctx, 250*time.Millisecond)
+			if next == len(entries) {
+				// Done with entries. No longer trigger on waitCh.
+				waitCh = nil
+			} else {
+				waitCh = waitCtx.Done()
+			}
+
+			go func(entry E, testDone context.CancelFunc) {
+				defer testDone()
+				result, err := test(entry)
```

Comment:
Maybe not in this PR, `test` can accept a `ctx` object to it knows when to cancel the request. Because there might be resource leak here:

1. go func () { test(entries[0]); }
2. go func() { test(entries[1]); }
3. test(entries[0]) finished with `nil`-error, returns
4. If test(entries[1]) blocks, there are resource leaks

Reply:
It's strictly not necessary. I would just be forwarding the input context. If you are calling raceTests, you are also providing the context and function, so you can write the input function in a way that uses the input context.

That's exactly what I do in the stream_dialer.go code:

```go
	resolver, err := raceTests(ctx, 250*time.Millisecond, resolvers, func(resolver *smartResolver) (*smartResolver, error) {
		for _, testDomain := range testDomains {
			select {
			case <-ctx.Done():
				return nil, ctx.Err()
			default:
			}
...
```